File: scripts/import_playground_cases.py
"""Admit reviewed browser cases into the shared program IR when their trace is exactly representable.

This adapter accepts natural-language leaf functions with successful root actions. It records
rejections for cases requiring richer graph/effect semantics instead of silently losing them.
"""
from __future__ import annotations

import argparse
import json
import sys
import tempfile
from pathlib import Path, PurePosixPath

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from natlang.codebase import load_function
from program_ir import VERSION, validate


def _safe_path(path: str) -> PurePosixPath:
    candidate = PurePosixPath(path)
    if candidate.is_absolute() or not candidate.parts or ".." in candidate.parts or candidate.suffix not in {".nl", ".ts"}:
        raise ValueError(f"unsafe or unsupported source path: {path}")
    return candidate


def _root(case: dict, folder: Path) -> dict:
    source = case.get("source") or {}
    files = source.get("files") or {}
    root_path = _safe_path(source.get("root", ""))
    if root_path.suffix != ".nl" or root_path.as_posix() not in files:
        raise ValueError("only natural-language root functions can become lambda scenarios")
    for name, body in files.items():
        path = _safe_path(name)
        if not isinstance(body, str):
            raise ValueError("source file must contain text")
        target = folder.joinpath(*path.parts)
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_text(body)
    function = load_function(folder.joinpath(*root_path.parts))
    if function.kind != "instructions" or function.codebase:
        raise ValueError("linked child functions need a graph adapter")
    return {"$lambda": {**function.to_lambda_doc(), "args": case.get("inputs") or {}}}


def _operations(events: list[dict]) -> list[dict]:
    actions = [event for event in events if event.get("kind") == "action"]
    if not actions:
        raise ValueError("no recorded actions")
    operations = []
    for action in actions:
        if action.get("call_id") != "$root@1" or action.get("outcome") != "ok":
            raise ValueError("nested or rejected actions need a richer adapter")
        args = action.get("arguments") or {}
        name = action.get("name")
        if name == "write" and isinstance(args.get("path"), str) and isinstance(args.get("type"), str):
            op = {"op": "assign", "target": args["path"], "value_type": args["type"]}
            if "source" in args:
                op["from"] = args["source"]
            elif "value" in args:
                op["value"] = args["value"]
            else:
                raise ValueError("write action has neither source nor value")
            if "done" in args:
                op["completion"] = args["done"]
            operations.append(op)
        elif name == "report_error" and isinstance(args.get("message"), str):
            operations.append({"op": "fail", "message": args["message"]})
        elif name == "report_blocker" and isinstance(args.get("missing"), str):
            operations.append({"op": "block", "missing": args["missing"]})
        else:
            raise ValueError(f"action {name!r} needs a richer adapter")
    return operations


def convert(case: dict, folder: Path) -> dict:
    if case.get("schema") != "natlang.playground.case/1" or case.get("reviewStatus") != "accepted" or case.get("admission", {}).get("admitted") is not True:
        raise ValueError("case is not reviewed and exactly admitted")
    events = case.get("trace") or []
    if not events or events[0].get("kind") != "manifest" or any(
        event.get("seq") != index or event.get("version") != "reduction-trace/1"
        for index, event in enumerate(events)
    ):
        raise ValueError("missing or discontinuous reduction trace")
    final = [event for event in events if event.get("kind") == "state" and event.get("phase") == "final"]
    if len(final) != 1:
        raise ValueError("trace needs one final state")
    expected = case.get("expected") or {}
    if expected.get("kind") != final[0].get("outcome") or (
        expected.get("kind") == "done" and expected.get("value") != final[0].get("value")
    ):
        raise ValueError("expected result differs from recorded outcome")
    if any(event.get("kind") == "effect" for event in events):
        raise ValueError("host effects need a richer adapter")
    root = _root(case, folder)
    operations = _operations(events)
    semantic_actions = [event for event in events if event.get("kind") == "action"]
    contract = {"kind": expected["kind"], "value": expected.get("value"),
                "effects": case.get("effects") or [],
                "required_actions": [{"tool": event["name"], "arguments": event["arguments"]}
                                     for event in semantic_actions],
                "instruction_obligations": case.get("requiredActions") or [],
                "constrained_calls": case.get("constrainedCalls") or []}
    if expected["kind"] in {"error", "blocked"}:
        last = semantic_actions[-1]
        contract["explanation"] = next(iter((last.get("arguments") or {}).values()), "")
    record = {"version": VERSION, "id": str(case["id"]), "kind": "lambda_scenario",
              "source": "browser_playground", "split": case.get("split", "train"),
              "source_ids": [str(case["id"])],
              "source_groups": [str(case.get("groupId") or case.get("projectId") or case["id"])],
              "source_revisions": [str(case["revision"])], "license": "user-authored",
              "gold_sources": ["reviewed-browser-case"],
              "semantics": {"root": root, "operations": operations, "contract": contract}}
    return validate(record)


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("src", type=Path)
    parser.add_argument("dst", type=Path)
    parser.add_argument("--rejects", type=Path)
    args = parser.parse_args()
    if args.src.resolve() == args.dst.resolve():
        parser.error("input and output must differ")
    args.dst.parent.mkdir(parents=True, exist_ok=True)
    rejects_path = args.rejects or args.dst.with_name(args.dst.stem + "-rejects.jsonl")
    accepted = rejected = 0
    with args.src.open() as source, args.dst.open("w") as output, rejects_path.open("w") as rejects:
        for line_number, line in enumerate(source, 1):
            if not line.strip():
                continue
            case = None
            try:
                case = json.loads(line)
                with tempfile.TemporaryDirectory(prefix="natlang-case-") as directory:
                    record = convert(case, Path(directory))
                output.write(json.dumps(record, ensure_ascii=False) + "\n")
                accepted += 1
            except Exception as error:
                rejects.write(json.dumps({"line": line_number, "id": case.get("id") if isinstance(
                    case, dict) else None, "reason": str(error)}, ensure_ascii=False) + "\n")
                rejected += 1
    print(json.dumps({"accepted": accepted, "rejected": rejected,
                      "program_ir": str(args.dst), "rejects": str(rejects_path)}))


if __name__ == "__main__":
    main()

File: ts-host/scripts/import-playground-cases.mjs
#!/usr/bin/env node
/** Convert reviewed, admitted playground cases into program IR: the project, its inputs, and the checked result. */
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { basename, dirname, extname, isAbsolute, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { PROGRAM_VERSION, programDefinition } from '../dist/teacher/program.js';

function checkSourcePath(value) {
  if (typeof value !== 'string' || !value || value.includes('\\') || value.includes('\0'))
    throw new Error(`unsafe or unsupported source path: ${value}`);
  const parts = value.split('/');
  if (isAbsolute(value) || parts.some(part => !part || part === '.' || part === '..') || !['.nl', '.ts'].includes(extname(value)))
    throw new Error(`unsafe or unsupported source path: ${value}`);
}

export async function convertCase(caseRecord) {
  if (caseRecord?.schema !== 'natlang.playground.case/1' || caseRecord.reviewStatus !== 'accepted' ||
      caseRecord.admission?.admitted !== true) throw new Error('case is not reviewed and exactly admitted');
  const source = caseRecord.source ?? {}, files = source.files ?? {};
  for (const [path, text] of Object.entries(files)) {
    checkSourcePath(path);
    if (typeof text !== 'string') throw new Error('source file must contain text');
  }
  if (extname(source.root ?? '') !== '.nl' || typeof files[source.root] !== 'string')
    throw new Error('only natural-language root functions can become program IR');
  const events = caseRecord.trace ?? [];
  if (!events.length || events[0].kind !== 'manifest' || events.some((event, index) =>
    event.seq !== index || event.version !== 'reduction-trace/1'))
    throw new Error('missing or discontinuous reduction trace');
  if (events.some(event => event.kind === 'effect')) throw new Error('host effects need an effect contract');
  const expected = caseRecord.expected ?? {};
  if (!['done', 'blocked'].includes(expected.kind)) throw new Error(`unsupported expected outcome: ${expected.kind}`);
  const id = String(caseRecord.id);
  const record = { version: PROGRAM_VERSION, id, kind: 'lambda_source', source: 'browser_playground',
    split: caseRecord.split ?? 'train', source_ids: [id],
    source_groups: [String(caseRecord.groupId ?? caseRecord.projectId ?? id)],
    source_revisions: [String(caseRecord.revision)], license: 'user-authored', gold_sources: ['reviewed-browser-case'],
    semantics: { root: source.root, files: structuredClone(files), inputs: structuredClone(caseRecord.inputs ?? {}),
      expected: expected.kind === 'done' ? expected.value : null, operation: expected.kind === 'blocked' ? 'blocked' : 'exact',
      contract: { required_actions: events.filter(event => event.kind === 'action').map(event => ({ tool: event.name, arguments: event.arguments })),
        instruction_obligations: caseRecord.requiredActions ?? [] } } };
  // The project must load: its root, callable folder, and declared inputs.
  const definition = programDefinition(record);
  for (const name of Object.keys(record.semantics.inputs))
    if (!definition.params.some(param => param.name === name)) throw new Error(`${name} is not a parameter of ${definition.name}`);
  return record;
}

export async function importCases(sourcePath, outputPath, rejectsPath = null) {
  const source = resolve(sourcePath), output = resolve(outputPath);
  if (source === output) throw new Error('input and output must differ');
  const rejectFile = resolve(rejectsPath ?? join(dirname(output), `${basename(output, extname(output))}-rejects.jsonl`));
  await mkdir(dirname(output), { recursive: true });
  const accepted = [], rejected = [];
  const lines = (await readFile(source, 'utf8')).split(/\r?\n/);
  for (let index = 0; index < lines.length; index++) {
    if (!lines[index].trim()) continue;
    let record = null;
    try {
      record = JSON.parse(lines[index]);
      accepted.push(await convertCase(record));
    } catch (error) {
      rejected.push({ line: index + 1, id: record && typeof record === 'object' ? record.id ?? null : null,
        reason: error instanceof Error ? error.message : String(error) });
    }
  }
  await writeFile(output, accepted.map(row => JSON.stringify(row)).join('\n') + (accepted.length ? '\n' : ''));
  await writeFile(rejectFile, rejected.map(row => JSON.stringify(row)).join('\n') + (rejected.length ? '\n' : ''));
  console.log(JSON.stringify({ accepted: accepted.length, rejected: rejected.length,
    program_ir: output, rejects: rejectFile }));
  return { accepted: accepted.length, rejected: rejected.length, output, rejects: rejectFile };
}

if (process.argv[1] && resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  const [, , source, output, ...rest] = process.argv;
  if (!source || !output || (rest.length && (rest.length !== 2 || rest[0] !== '--rejects'))) {
    console.error('usage: node scripts/import-playground-cases.mjs INPUT.jsonl OUTPUT.jsonl [--rejects FILE]');
    process.exitCode = 2;
  } else {
    const rejectAt = rest[0] === '--rejects' ? process.argv.indexOf('--rejects') + 1 : -1;
    const rejects = rejectAt > 0 ? process.argv[rejectAt] : null;
    if (rest.length && !rejects) { console.error('unknown option'); process.exitCode = 2; }
    else importCases(source, output, rejects).catch(error => {
      console.error(error instanceof Error ? error.stack : String(error)); process.exitCode = 1;
    });
  }
}
